### PR TITLE
[SPARK-27347][MESOS] Fix supervised driver retry logic for outdated tasks

### PR DESCRIPTION
**Pull request by samvantran**
## What changes were proposed in this pull request?

This patch fixes a bug where `--supervised` Spark jobs would retry multiple times whenever an agent would crash, come back, and re-register even when those jobs had already relaunched on a different agent.

That is: 
```
- supervised driver is running on agent1
- agent1 crashes
- driver is relaunched on another agent as `<task-id>-retry-1`
- agent1 comes back online and re-registers with scheduler
- spark relaunches the same job as `<task-id>-retry-2`
- now there are two jobs running simultaneously
```

This is because when an agent would come back and re-register it would send a status update `TASK_FAILED` for its old driver-task. Previous logic would indiscriminately remove the `submissionId` from Zookeeper's `launchedDrivers` node and add it to `retryList` node. Then, when a new offer came in, it would relaunch another `-retry-`  task even though one was previously running. 

For example logs, scroll to bottom

## How was this patch tested?

- Added a unit test to simulate behavior described above
- Tested manually on a DC/OS cluster by
  ```
  - launching a --supervised spark job
  - dcos node ssh <to the agent with the running spark-driver>
  - systemctl stop dcos-mesos-slave
  - docker kill <driver-container-id>
  - [ wait until spark job is relaunched ]
  - systemctl start dcos-mesos-slave
  - [ observe spark driver is not relaunched as `-retry-2` ]
  ```

Log snippets included below. Notice the `-retry-1` task is running when status update for the old task comes in afterward: 
```
19/01/15 19:21:38 TRACE MesosClusterScheduler: Received offers from Mesos: 
... [offers] ...
19/01/15 19:21:39 TRACE MesosClusterScheduler: Using offer 5d421001-0630-4214-9ecb-d5838a2ec149-O2532 to launch driver driver-20190115192138-0001 with taskId: value: "driver-20190115192138-0001"
...
19/01/15 19:21:42 INFO MesosClusterScheduler: Received status update: taskId=driver-20190115192138-0001 state=TASK_STARTING message=''
19/01/15 19:21:43 INFO MesosClusterScheduler: Received status update: taskId=driver-20190115192138-0001 state=TASK_RUNNING message=''
...
19/01/15 19:29:12 INFO MesosClusterScheduler: Received status update: taskId=driver-20190115192138-0001 state=TASK_LOST message='health check timed out' reason=REASON_SLAVE_REMOVED
...
19/01/15 19:31:12 TRACE MesosClusterScheduler: Using offer 5d421001-0630-4214-9ecb-d5838a2ec149-O2681 to launch driver driver-20190115192138-0001 with taskId: value: "driver-20190115192138-0001-retry-1"
...
19/01/15 19:31:15 INFO MesosClusterScheduler: Received status update: taskId=driver-20190115192138-0001-retry-1 state=TASK_STARTING message=''
19/01/15 19:31:16 INFO MesosClusterScheduler: Received status update: taskId=driver-20190115192138-0001-retry-1 state=TASK_RUNNING message=''
...
19/01/15 19:33:45 INFO MesosClusterScheduler: Received status update: taskId=driver-20190115192138-0001 state=TASK_FAILED message='Unreachable agent re-reregistered'
...
19/01/15 19:33:45 INFO MesosClusterScheduler: Received status update: taskId=driver-20190115192138-0001 state=TASK_FAILED message='Abnormal executor termination: unknown container' reason=REASON_EXECUTOR_TERMINATED
19/01/15 19:33:45 ERROR MesosClusterScheduler: Unable to find driver with driver-20190115192138-0001 in status update
...
19/01/15 19:33:47 TRACE MesosClusterScheduler: Using offer 5d421001-0630-4214-9ecb-d5838a2ec149-O2729 to launch driver driver-20190115192138-0001 with taskId: value: "driver-20190115192138-0001-retry-2"
...
19/01/15 19:33:50 INFO MesosClusterScheduler: Received status update: taskId=driver-20190115192138-0001-retry-2 state=TASK_STARTING message=''
19/01/15 19:33:51 INFO MesosClusterScheduler: Received status update: taskId=driver-20190115192138-0001-retry-2 state=TASK_RUNNING message=''
```

### CODE REVIEW

**Comment by samvantran** [comment]
cc @vanzin @tnachen @susanxhuynh @mgummelt because

File: resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala [hunk 1]
```diff
@@ -786,6 +790,15 @@ private[spark] class MesosClusterScheduler(
...
   }
 
+  /**
+   * Check if the task has already been launched or is pending
+   * If neither, the taskId is outdated and should be ignored
+   * This is to avoid scenarios where an outdated status update arrives
+   * after a supervised driver has already been relaunched
+   */
+  private def taskIsOutdated(taskId: String, state: MesosClusterSubmissionState): Boolean =
+    taskId != state.taskId.getValue && !pendingRetryDrivers.contains(state.driverDescription)
```
Comment (susanxhuynh) [inline resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala]: Does this include the case where the task was launched a while ago and has already finished? (Is that even a valid case?) Would this still return "True"?

Comment (samvantran) [reply resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala]: Yup. So in your example say: 
```
- driver task01 running on agent1
- agent1 crashes
- job relaunched on agent2 as task01-retry-1
- job finishes. Latest state.task.Id == task01-retry-1
- agent1 comes back and tries to relaunch task01
- this method would see that task01 =/= task01-retry-1 && 
  there aren't any pendingRetries and would return True
  preventing a zombie job from reviving
```

**Comment by samvantran** [comment]
light ping @vanzin @tnachen

**Comment by dongjoon-hyun** [comment]
ok to test

File: resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala [hunk 2]
```diff
@@ -786,6 +790,15 @@ private[spark] class MesosClusterScheduler(
     }
   }
 
+  /**
+   * Check if the task has already been launched or is pending
+   * If neither, the taskId is outdated and should be ignored
+   * This is to avoid scenarios where an outdated status update arrives
+   * after a supervised driver has already been relaunched
+   */
+  private def taskIsOutdated(taskId: String, state: MesosClusterSubmissionState): Boolean =
```
Comment (dongjoon-hyun) [inline resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala]: `taskIsOutdated` -> `isTaskOutdated`?

File: resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala [hunk 3]
```diff
@@ -345,7 +345,7 @@ private[spark] class MesosClusterScheduler(
       this.masterInfo = Some(masterInfo)
       this.schedulerDriver = driver
 
-      if (!pendingRecover.isEmpty) {
+      if (pendingRecover.nonEmpty) {
```
Comment (dongjoon-hyun) [inline resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala]: I understand what this is, but could you revert this unrelated change?

File: resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala [hunk 4]
```diff
@@ -766,6 +766,10 @@ private[spark] class MesosClusterScheduler(
         val state = launchedDrivers(subId)
         // Check if the driver is supervise enabled and can be relaunched.
         if (state.driverDescription.supervise && shouldRelaunch(status.getState)) {
+          if (taskIsOutdated(taskId, state)) {
+            // Prevent outdated task from overwriting a more recent status
+            return
```
Comment (dongjoon-hyun) [inline resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala]: @samvantran . `shouldRelaunch` looks like more natural place to check this kind of logic. How do you think about this? Do we need `taskIsOutdated` logic? This is only used here.

File: resource-managers/mesos/src/test/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterSchedulerSuite.scala [hunk 5]
```diff
@@ -450,6 +450,84 @@ class MesosClusterSchedulerSuite extends SparkFunSuite with LocalSparkContext wi
     assert(state.finishedDrivers.size == 1)
   }
 
+  test("does not restart outdated supervised drivers") {
```
Comment (dongjoon-hyun) [inline resource-managers/mesos/src/test/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterSchedulerSuite.scala]: Can we rename the test case?
```scala
- test("does not restart outdated supervised drivers") {
+ test("SPARK-27347 do not restart outdated supervised drivers") {
```

File: resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala (hunk 3 above)
Comment (samvantran) [reply resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala]: done!

File: resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala (hunk 2 above)
Comment (samvantran) [reply resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala]: thanks!

File: resource-managers/mesos/src/test/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterSchedulerSuite.scala (hunk 5 above)
Comment (samvantran) [reply resource-managers/mesos/src/test/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterSchedulerSuite.scala]: happy to

File: resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala (hunk 4 above)
Comment (samvantran) [reply resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala]: Thanks for the review @dongjoon-hyun. I originally thought so too but this doesn't work because if we include the `taskIsOutdated` logic in the `shouldRelaunch` method, we end up skipping up-to-date task statuses and not adding them to the `pendingRetryDrivers`.  

Put another way, we actually do need to separately evaluate if a task is
1. supervised and `TASK_LOST` / `TASK_FAILED`
2. and whether it is valid (and should process the logic of L773-782) or skip it because we've seen the id and have already done this once before.

If we don't include `taskIsOutdated` logic here, then we end up incorrectly `removingLaunchedDrivers`, even though the job was relaunched on another agent, and adding a duplicate job to `pendingRetryDrivers` even though it shouldn't be retried.

**Review by samvantran — COMMENTED** [review]
Thank you for the review @dongjoon-hyun! I've made changes to the PR based on your comments. Please take a 2nd look.

**Comment by dongjoon-hyun** [comment]
Retest this please.

File: resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala [hunk 6]
```diff
@@ -786,6 +790,15 @@ private[spark] class MesosClusterScheduler(
...
   }
 
+  /**
+   * Check if the task is outdated i.e. has already been launched or is pending
+   * If neither, the taskId is outdated and should be ignored
+   * This is to avoid scenarios where an outdated status update arrives
+   * after a supervised driver has already been relaunched
+   */
+  private def isTaskOutdated(taskId: String, state: MesosClusterSubmissionState): Boolean =
+    taskId != state.taskId.getValue && !pendingRetryDrivers.contains(state.driverDescription)
```
Comment (dongjoon-hyun) [inline resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala]: @samvantran . This works, but `pendingRetryDrivers` are assumed to be keyed by submission id. We should use `pendingRetryDrivers.exists(_.submissionId == state.driverDescription.submissionId)` instead of `pendingRetryDrivers.contains` to be robust. Could you fix like that?

File: resource-managers/mesos/src/test/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterSchedulerSuite.scala [hunk 7]
```diff
@@ -450,6 +450,84 @@ class MesosClusterSchedulerSuite extends SparkFunSuite with LocalSparkContext wi
...
+      .setTaskId(TaskID.newBuilder().setValue(response.submissionId).build())
+      .setSlaveId(agent1)
+      .setReason(TaskStatus.Reason.REASON_SLAVE_REMOVED)
+      .setState(MesosTaskState.TASK_LOST)
+      .build()
+
+    scheduler.statusUpdate(driver, taskStatus)
+    state = scheduler.getSchedulerState()
+    assert(state.pendingRetryDrivers.size == 1)
+    assert(state.launchedDrivers.isEmpty)
```
Comment (dongjoon-hyun) [inline resource-managers/mesos/src/test/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterSchedulerSuite.scala]: If we check here, I guess `!pendingRetryDrivers.contains(state.driverDescription)` logic might fail because `retryState` might be different.

BTW, could you improve the test case to verify both `taskId != state.taskId.getValue` and `!pendingRetryDrivers.contains(state.driverDescription)` explicitly?

**Comment by dongjoon-hyun** [comment]
Ping, @samvantran .

File: resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala (hunk 6 above)
Comment (samvantran) [reply resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala]: Updated!

File: resource-managers/mesos/src/test/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterSchedulerSuite.scala (hunk 7 above)
Comment (samvantran) [reply resource-managers/mesos/src/test/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterSchedulerSuite.scala]: Hm, in the test `state` is a `MesosClusterSchedulerState` object and does not have `taskId` nor `driverDescription` fields (that is in `MesosClusterSubmissionState`). 

However at this point, `state.pendingRetryDrivers` **does** contain the same `taskStatus.taskId` (ie `sub1` in the test). That's because the task failed, got added to `pending` and is awaiting the next suitable offer. Later below on L506, a new offer comes in and the task is moved out of `pending` and into `launched` (with a new taskId: `sub1-retry-1`). From then on, what I'm testing is any new updates with the old taskId (`sub1`) have no effect. 

But I've added two more tests that are more explicit about what IDs match where. Is that all right?

**Review by samvantran — COMMENTED** [review]
Thanks @dongjoon-hyun. I've applied your change with a response to your test questions.

File: resource-managers/mesos/src/test/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterSchedulerSuite.scala (hunk 7 above)
Comment (dongjoon-hyun) [reply resource-managers/mesos/src/test/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterSchedulerSuite.scala]: Thanks!

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Merged to master/branch-2.4/branch-2.3.
Thank you for your first contribution.

**Comment by dongjoon-hyun** [comment]
@samvantran . I added you into an Apache Spark contributor group and assigned the issue to you.
Welcome to the Apache Spark community!
